src/Core/Algorithms/Legacy/Fields/Cleanup/ReorderNormalCoherentlyAlgo.cc

```diff
@@ -0,0 +1,273 @@
+   DEALINGS IN THE SOFTWARE.
+   */
+
+#include <Core/Algorithms/Legacy/Fields/Cleanup/ReorderNormalCoherentlyAlgo.h>
+#include <Core/Datatypes/Legacy/Field/Field.h>
+#include <Core/Datatypes/Legacy/Field/VField.h>
+#include <Core/Datatypes/Matrix.h>
+#include <Core/Datatypes/MatrixTypeConversions.h>
+#include <Core/Datatypes/SparseRowMatrix.h>
+#include <Core/Datatypes/DatatypeFwd.h>
+#include <Core/Datatypes/Legacy/Field/FieldInformation.h>
+#include<Core/Datatypes/Legacy/Field/VMesh.h>
+#include <Core/Datatypes/DenseColumnMatrix.h>
+#include <Core/Datatypes/PropertyManagerExtensions.h>
+
+#include <Core/Algorithms/Base/AlgorithmVariableNames.h>
+#include <Core/Algorithms/Base/AlgorithmPreconditions.h>
+
+#include<queue>
+
+using namespace SCIRun::Core::Algorithms;
+using namespace SCIRun::Core::Algorithms::Fields;
+using namespace SCIRun::Core::Algorithms::Fields::Parameters;
+using namespace SCIRun::Core::Datatypes;
+using namespace SCIRun;
+
+ALGORITHM_PARAMETER_DEF(Fields, invertedElementsCheckBox);
+
+ReorderNormalCoherentlyAlgo::ReorderNormalCoherentlyAlgo()
+{
+  addParameter(Parameters::invertedElementsCheckBox, false);
+}
+
+void ReorderNormalCoherentlyAlgo::runImpl(FieldHandle inputField, FieldHandle& outputField, DenseColumnMatrixHandle& invertedElementsListMatrix) const
+{
+  if(!inputField)
+  {
+    THROW_ALGORITHM_INPUT_ERROR("No input field");
+  }
+
+  FieldInformation fi(inputField);
+  
+  if(fi.is_nonlinear())
+    error("The algorithm has not yet been defined for non-linear elements yet");
+  
+  if(!fi.is_trisurfmesh())
+    error("This algorithm only works on a TriSurfMesh");
+  
+  VField* inputVField=inputField->vfield();
+  VMesh* inputVMesh=inputField->vmesh();
+  
+  fi.make_trisurfmesh();
+  outputField=CreateField(fi);
+  
+  if(!outputField)
+    error("Could not allocate output field");
+  
+  VMesh *outputVMesh=outputField->vmesh();
+  const VMesh::size_type n=inputVMesh->num_nodes();
+  const VMesh::size_type m=inputVMesh->num_elems();
+  outputVMesh->node_reserve(n);
+  outputVMesh->elem_reserve(m);
+  
+  outputVMesh->copy_nodes(inputVMesh);
+  
+  size_type i,j,k;
+  std::vector<int> invertedElements;
+  
+  std::vector<std::set<int>> elemsOfVert;
+  std::vector<std::set<int>> elemNeighbors;
+  std::vector<std::vector<int>> elem;
+  std::set<int> dummy;
+  std::vector<int> tempVec;
+  
+  for(i=0;i<3;i++)
+    tempVec.push_back(0);
+  
+  //intialize vector of vectors
+  for(i=0;i<m;i++)
+    elem.push_back(tempVec);
+  
+  VMesh::Face::iterator meshFaceIter;
+  VMesh::Face::iterator meshFaceEnd;
+  
+  VMesh::Node::array_type nodesFromFace(3);
+  
+  inputVMesh->synchronize(Mesh::ALL_ELEMENTS_E);
+  
+  inputVMesh->end(meshFaceEnd);
+  
+  for(inputVMesh->begin(meshFaceIter); meshFaceIter!=meshFaceEnd; ++meshFaceIter)
+  {
+    // get nodes from mesh element
+    VMesh::Face::index_type elemID=*meshFaceIter;
+    inputVMesh->get_nodes(nodesFromFace, elemID);
+    
+    for(i=0;i<3;i++)
```

Comment:
[Magic number](https://refactoring.guru/replace-magic-number-with-symbolic-constant) `3` is used a lot, please extract a const variable for it.